Colr v1 api graph

drott opened the pull request.

drott committed:
Parse COLR v1 table header and offset to v1 base glyph array

drott committed:
wip type definitions

drott committed:
initial round of colr definitions

drott committed:
Add definition for retrieving color stops

drott committed:
Add definition for retrieving color stops

drott committed:
gradients extraction API piped through to ttcolr impl

drott committed:
Reading glyph ids v1 style

drott committed:
safety check

drott committed:
WIP reading layer length

drott committed:
Read 16bit glyph ids for now

drott committed:
Add copyright

drott committed:
Read points of linear gradients

drott committed:
Basic bounds check for paint format

drott committed:
colorstop argument to retrieving color stops

drott committed:
Support for retrieving color stops

drott committed:
Reading solid fills

drott committed:
radial extraction

drott committed:
Formatting

drott committed:
reorder radial struct, remove gid TODOs

drott committed:
Initial documentation for the freetype API extension

drott committed:
Clean up reformatting in freetype.h

drott committed:
Cleanup sfnt.h formatting changes

drott committed:
Documentation naming.

drott committed:
Union -> struct, remove format members

drott committed:
read affine updates

drott committed:
Work in progress, define external interface

drott committed:
WIP: OpaquePaint definition, GetPaint function, additional Paint structs

drott committed:
WIP, trying to read v1 DAG layers

drott committed:
implement parsing of additional paint formats

drott committed:
Fix COLR glyph size, offsets

drott committed:
Merged U suffix from upstream, fixed include path

drott commented on include/freetype/freetype.h:
```diff
@@ -4176,6 +4176,457 @@ FT_BEGIN_HEADER
+    FT_UInt glyphID;
+  } FT_PaintColrGlyph;
+
+  typedef struct FT_Affine_23_
```
Note to self, after discussion in https://github.com/googlefonts/colr-gradients-spec/pull/85, this needs to change to:

```
typedef struct FT_Affine_23_
  {
    FT_Fixed  xx, yx;
    FT_Fixed  xy, yy;
    FT_Fixed  dx, dy;

  } FT_Affine23;
```

drott replied:
Or alternatively row-major order, to be consistent with `FT_Matrix`:
```FT_Affine_23 as {xx, xy, dx, yx, yy, dy}```.

drott replied:
That's now done.

drott committed:
API change compiling

drott committed:
initializing iterator compiling

drott committed:
New API compiling

drott committed:
Fix first_layer_index related computations

drott committed:
Matrix sequence fixes

drott committed:
Fix layer pointer sanity check

lemzwerg commented on include/freetype/freetype.h:
```diff
@@ -4296,8 +4319,7 @@ FT_BEGIN_HEADER
    */
-  FT_EXPORT( FT_UShort )
-  FT_Get_FSType_Flags( FT_Face  face );
+  FT_EXPORT (FT_UShort) FT_Get_FSType_Flags (FT_Face face);
```
Please undo this change – FreeType uses the first line for the return value.

Additionally, the change is unrelated.

lemzwerg commented on include/freetype/freetype.h:
```diff
@@ -4204,29 +4204,62 @@ FT_BEGIN_HEADER
-
-
-  typedef struct  FT_ColorStopIterator_
+  typedef struct FT_ColorStopIterator_
```
Another unnecessary change – FreeType intentionally uses two spaces after the `struct` keyword.  I know this is weird, and no formatter supports that, but...

And please use two empty lines after a `struct` block, a function block, etc., etc.  Or rather, don't change the formatting of already present code! I will take care of cleaning up your code in a follow-up step to take away this burden from you.

lemzwerg commented on include/freetype/internal/sfnt.h:
```diff
@@ -555,12 +565,10 @@ FT_BEGIN_HEADER
-                         FT_GlyphSlot  base_glyph,
-                         FT_GlyphSlot  new_glyph );
-
+  typedef FT_Error ( *TT_Blend_Colr_Func ) ( TT_Face      face,
```
Loooots of unnecessary reformatting here and below...

lemzwerg commented on src/base/ftobjs.c:
```diff
@@ -5566,5 +5566,54 @@
+                                       FT_LayerIterator *iterator )
+  {
+    TT_Face       ttface;
+    SFNT_Service  sfnt;
```
FreeType uses two empty lines after the block declaration of variables.  Not an important thing for you, but I just want to mention it :-)

lemzwerg commented on src/sfnt/ttcolr.c:
```diff
@@ -352,7 +353,8 @@
         return 0;
 
-      p = (FT_Byte *)( colr->table + base_glyph_v1_record.layer_array_offset );
+      p                    = (FT_Byte *)( colr->base_glyphs_v1 +
```
Strange formatting...

lemzwerg commented on include/freetype/freetype.h:
```diff
@@ -4223,7 +4223,7 @@ FT_BEGIN_HEADER
   {
     FT_UInt16  palette_index;
-    FT_F2Dot14 transparency;
+    FT_F2Dot14 alpha;
```
This change is not reflected in the commit message...

lemzwerg commented on include/freetype/internal/sfnt.h:
```diff
@@ -531,9 +531,7 @@ FT_BEGIN_HEADER
-      TT_Face               face,
-      FT_PaintColor *       color_stop,
-      FT_ColorStopIterator *iterator );
+      TT_Face face, FT_ColorStop *color_stop, FT_ColorStopIterator *iterator );
```
Uuh, really strange formatting :-)

lemzwerg commented on src/sfnt/ttcolr.c:
```diff
@@ -365,7 +385,55 @@
       FT_NEXT_ULONG ( p );
       apaint->u.linear_gradient.p2.y = FT_NEXT_SHORT ( p );
       FT_NEXT_ULONG ( p );
+    } else if ( apaint->format == COLR_PAINTFORMAT_RADIAL_GRADIENT )
```
In FreeType, `else` (or `else if`) should be on a new line.

lemzwerg commented on include/freetype/freetype.h:
```diff
@@ -4213,6 +4223,26 @@ FT_BEGIN_HEADER
+   *
+   * @fields:
+   *   num_colors ::
+   *     The number of color stops for the requested glyph index.  Will be
```
Please don't use future tense.

lemzwerg commented on include/freetype/freetype.h:
```diff
@@ -4221,37 +4251,145 @@ FT_BEGIN_HEADER
+   *     The palette index into a CPAL palette.
+   *
+   *   alpha ::
+   *     Alpha transparency value applied on top of the value from CPAL
```
Missing full stop at end of sentence.

lemzwerg commented on src/sfnt/ttcolr.c:
```diff
@@ -345,22 +345,32 @@
-               FT_Byte *      layer_v1_array,
-               FT_ULong       paint_offset,
-               FT_COLR_Paint *apaint )
+  read_paint( Colr* colr, FT_Byte* p, FT_COLR_Paint* apaint )
```
I *hate* the clang formatter for doing this!  There is no possibility to configure it to do one variable per line even if the line is short.  Needless to say that FreeType needs one line per variable...

lemzwerg reviewed (commented):
I walked over commit by commit and added some comments, mainly related to formatting.

For a FreeType submission I would like to have the final changes split into small, logical chunks (that ideally do compile each), together with ChangeLog entries, which should be identical to commit messages (sans the leading tab character).

Don't worry too much about the formatting; I will take care of that :-)

Besides that, the code looks good, thanks!

drott committed:
Implement PaintRotate and PaintSkew

lemzwerg committed:
.mailmap: Add Priyesh.

lemzwerg committed:
[type42] Pacify static analysis tools (#59682).

* src/type42/t42objs.c (T42_Size_Init, T42_GlyphSlot_Init): Avoid
warnings about uninitialized variables.

drott committed:
Documentation polish

apodtele committed:
Documentation edits.

chris-liddell committed:
[truetype] Fix incremental metrics (#59503).

* src/truetype/ttgload.c (tt_get_metrics, load_truetype_glyph):
Previously, the code would populate the phantom points before
calling the `get_glyph_metrics` callback.  For formats like PCL XL
format 1, class 2 downloaded fonts (where metrics are removed from
the TTF header), this causes problems when the hinting program uses
the phantom points (misplaced and distorted glyphs) due to the
metrics being unset (all zeros).
(tt_get_metrics_incr_overrides): Renamed to...
(tt_get_metrics_incremental): ... this.  Updated caller

* include/freetype/ftincrem.h: Update the documentation to make it
clearer that `get_glyph_metrics` is to retrieve metrics from a
non-standard source, but *not* for the purpose of imposing custom
metrics.

drott committed:
Reformatting, reducing diff

drott commented:
Thanks for the initial review, Werner. In a first pass, I addressed the formatting comments. I'll work on splitting this into separate PRs now.

drott committed:
Rename FT_PaintColor to FT_ColorIndex

drott committed:
Implement FT_PaintTranslate

drott committed:
two newlines after block declarations
